### ISSUE TITLE
Example of generic IRequestHandler

### ISSUE THREAD

**Issue by cirrusone** [jbogard/MediatR#521]
Are there any examples to show how to make a generic IRequestHandler?

I can obviously add lots of interfaces such as
```
public class MainWindowViewModel : BindableBase
,IRequestHandler<MainViewModelPing1, string>
,IRequestHandler<MainViewModelPing2, string>
,IRequestHandler<MainViewModelPing3, SomeOtherClass>
```

but cannot find any examples of a generic handler to handle different requests and responses in the same handler. Is something like the following possible?

```
public class MainViewModelPing : IRequest<T>
    {
        public string Message { get; set; }
	public T Data { get; set; }
    }
```

which would handle something like the following

```
    public class MainViewModelPing : IRequest<string>
    {
        public string Message { get; set; }
    }

    public class MainViewModelPing : IRequest<string>
    {
        public string Message { get; set; }
	public SomeClass Data { get; set; }
    }

    public class MainViewModelPing : IRequest<SomeOtherClass>
    {
        public string Message { get; set; }
	public SomeClass Data { get; set; }
    }
```

**Comment by cirrusone** [comment]
Is this a reasonable way to use Mediatr without any foreseeable issues?

App DI:
```
protected override void RegisterTypes(IContainerRegistry containerRegistry)
{
    // MediatR DI dryloc
    var container = containerRegistry.GetContainer();
    container.RegisterDelegate<ServiceFactory>(r => r.Resolve);
    container.RegisterMany(new[] { typeof(IMediator).GetAssembly(), typeof(MainViewModelPing).GetAssembly() }, Registrator.Interfaces);

}
```

GenericRequest and GenericResponse:
```
public class MainViewModelPing : IRequest<PingResponse>
{
  public PingRequest Request { get; set; }
}

public abstract class PingRequest
{
}
public class GenericRequest<T> : PingRequest
{
  public string Message { get; set; }
  public T Data { get; set; }
}

public abstract class PingResponse
{
}
public class GenericResponse<T> : PingResponse
{
  public string Message { get; set; }
  public T Data { get; set; }
}
```

Generic Handler attached to MainViewModel:
```
public class MainWindowViewModel : BindableBase, IRequestHandler<MainViewModelPing, PingResponse>
{

        public Task<PingResponse> Handle(MainViewModelPing request, CancellationToken cancellationToken)
        {   
            // Test types and message to filter
            if(request.Request is GenericRequest<string> stringRequest)
            {
                if(stringRequest.Message.Equals("SomeVal"))
                {
                    // Perform action here
                }
            }
            else if (request.Request is GenericRequest<int> intRequest)
            {
                if(stringRequest.Message.Equals("SomeOtherVal"))
                {
                    // Perform action here
                }
            }
            
            // Example response
            var genericResponse = new GenericResponse<int> { Message = "It Works", Data = 42 };
            return Task.FromResult(genericResponse as PingResponse);
        }
}
```

Example request:
```
        private async void MediatrTest()
        {
            var genericRequest = new GenericRequest<string> { Message = "SomeVal", Data = "" };
            var response = await _mediator.Send(new MainViewModelPing() { Request = genericRequest });
        }
```

**Comment by no1melman** [comment]
I would say, making it more generic isn't an ideal use case.

The idea is that you have 1 request that maps to 1 handler because this follows SRP. It becomes easier to understand what the code is trying to achieve.

I would say that if you just want `T` to be stored in a database, handled the same way each time... I would probably just make it `object Data {get;set;}` and do some kind of reflection in the handler to get the object to "fit" into the database.

There maybe a design issue with what you're doing which has made this pattern emerge.

The pipeline handlers show how to "pass through" these requests and sort of do something with them.

**Comment by hanslai** [comment]
@cirrusone do you have this figured out.  I am trying to do the same thing with asp.net core DI container, have not find the solution yet.  Thanks

**Comment by cirrusone** [comment]
I actually changed to `Microsoft.Toolkit.Mvvm` as the messenger suited my use-case better. It also allows me to register lots of message types without lots of interfaces. _Since it targets .NET Standard 2.0, this means that it can be used anywhere from UWP apps, to Uno, Xamarin, Unity, ASP.NET, etc. Literally any framework supporting the .NET Standard 2.0 feature set. There are no platform specific dependencies at all. The whole package is entirely platform, runtime, and UI stack agnostic._

https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/3428

https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/3230

https://github.com/windows-toolkit/MVVM-Samples/blob/master/docs/mvvm/Messenger.md

**Comment by hanslai** [comment]
@cirrusone Thanks a lot, I will check it out also.

**Comment by agot11** [comment]
@zachpainter77 

https://github.com/jbogard/MediatR.Extensions.Microsoft.DependencyInjection

![image](https://user-images.githubusercontent.com/43707500/125126732-2f9eff00-e104-11eb-8e5c-cf625e8e844f.png)

**Comment by zachpainter77** [comment]
It is my understanding (and I tested this) that explicitly registering the GenericHandlerBase isn't necessary... If you are going to create concrete handlers for each generic type for that handler then MediatR will find the correct handler without issue.. 

Here is an example of this:
![image](https://user-images.githubusercontent.com/9939026/129097926-10320e03-495d-420f-bfc2-84a03bb9f169.png)
![image](https://user-images.githubusercontent.com/9939026/129097991-9bf541c2-ced4-465e-a9c5-46398abfe9fc.png)

Here is the registration using Default DI Container:
![image](https://user-images.githubusercontent.com/9939026/129098217-a9f07af1-7313-48da-a957-df2ac90f1383.png)

This works just fine... Now imagine if you had multiple entities, and further more, imagine if you had multiple service dependencies for your handler.  While even though you would not have to implement the handle method for each derived handler you still need to create a concrete class that closes the generic type and pass all dependencies to the base constructor... In my opinion that is very tedious to have to do that for every entity and every service dependency.. In my opinion it is much simpler to simply register the generic handler via Autofac like I showed above.  But if not using a third party DI container then by all means create the concrete handlers... If I couldn't use a third party di container for whatever reason then I would do it this way.. 

Cheers.
